src/librustc_front/lowering.rs

```diff
@@ -276,14 +284,22 @@ pub fn lower_variant(lctx: &LoweringContext, v: &Variant) -> P<hir::Variant> {
     })
 }
 
-pub fn lower_path(lctx: &LoweringContext, p: &Path) -> hir::Path {
+// Path segments are usually unhygienic, hygienic path segments can occur only in
+// identifier-like paths originating from `ExprPath`.
+// Make life simpler for rustc_resolve by renaming only such segments.
```

Comment:
It requires many uses of `unhygienic_name` in all kind of paths all over the resolve instead of one place in `resolve_identifier`. Most of paths are unhygienic, only identifiers are hygienic.
I even thought about splitting `ExprPath` into `ExprPath` and `ExprIdent` (by analogy with `PatEnum` and `PatIdent` in patterns) to make this problem go away.